pkg/kubelet/cm/devicemanager/pod_devices.go

```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 			annotationsMap[k] = v
 			opts.Annotations = append(opts.Annotations, kubecontainer.Annotation{Name: k, Value: v})
 		}
+
+		// Updates for CDI devices. This also updates the annotations.
+		cdiDevices, cdiAnnotations := getCDIDeviceInfo(
+			podUID, resourceName,
+			resp.CdiDevices,
+			annotationsMap,
+			cdiDevicesMap,
+		)
+		opts.CDIDevices = append(opts.CDIDevices, cdiDevices...)
+		opts.Annotations = append(opts.Annotations, cdiAnnotations...)
 	}
 	return opts
 }
 
+// getCDIDeviceInfo returns CDI devices and annotations for a given pod.
+func getCDIDeviceInfo(podUID, resourceName string, responseCDIDevices []*pluginapi.CDIDevice, annotationsMap map[string]string, cdiDevicesMap map[string]struct{}) ([]kubecontainer.CDIDevice, []kubecontainer.Annotation) {
+	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.DevicePluginCDIDevices) {
+		return nil, nil
+	}
+
+	var cdiDeviceNames []string
+	for _, cdiDevice := range responseCDIDevices {
+		if _, ok := cdiDevicesMap[cdiDevice.Name]; ok {
+			klog.V(4).InfoS("Skip existing CDI Device", "name", cdiDevice.Name)
+			continue
+		}
+		klog.V(4).InfoS("Add CDI device", "name", cdiDevice.Name)
+		cdiDevicesMap[cdiDevice.Name] = struct{}{}
+
+		cdiDeviceNames = append(cdiDeviceNames, cdiDevice.Name)
+	}
+
+	var cdiDevices []kubecontainer.CDIDevice
+	for _, deviceName := range cdiDeviceNames {
+		device := kubecontainer.CDIDevice{
+			Name: deviceName,
+		}
+		cdiDevices = append(cdiDevices, device)
+	}
+
+	annotationUID := strings.Join([]string{podUID, resourceName}, "-")
+	cdiAnnotations, err := cdi.GenerateCDIAnnotations(types.UID(annotationUID), "devicemanager", cdiDeviceNames)
+	if err != nil {
+		klog.ErrorS(err, "Failed to create CDI annotations")
+		return cdiDevices, nil
+	}
+
+	var annotations []kubecontainer.Annotation
+	for _, a := range cdiAnnotations {
+		if e, ok := annotationsMap[a.Name]; ok {
+			klog.V(4).InfoS("Skip existing annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+			if e != a.Value {
+				klog.ErrorS(nil, "Annotation has conflicting setting", "annotationKey", a.Name, "expected", e, "got", a.Value)
+			}
+			continue
+		}
+		klog.V(4).InfoS("Add annotation", "annotationKey", a.Name, "annotationValue", a.Value)
+		annotationsMap[a.Name] = a.Value
+		annotations = append(annotations, kubecontainer.Annotation{Name: a.Name, Value: a.Value})
+	}
```

Comment:
As mentioned previously, lets pull this out to its own function and called outside of this function.